### PR TITLE
Fixes #29910 - Refactor setting option selection

### PR DESCRIPTION
**Pull request by xprazak2**
There should be no changes in behaviour or visuals.

### CODE REVIEW

File: test/unit/setting_value_selection_test.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
+require 'test_helper'
+require 'pry'
```
Comment (timogoebel) [inline test/unit/setting_value_selection_test.rb]:
```suggestion
```

File: app/models/setting.rb [hunk 2]
```diff
@@ -287,8 +287,7 @@ def self.set(name, description, default, full_name = nil, value = nil, options =
     if options.has_key? :collection
       SettingsHelper.module_eval do
         define_method("#{name}_collection".to_sym) do
```
Comment (ezr-ondrej) [inline app/models/setting.rb]: Do you think we could have a registry on `SettingValueSelection`, that would keep a hash of `[settting_name] => options[:collection]`?
That way we could get rid of this dynamic method definition, in favor of something like `SettingValueSelection.has_collection?(name)` and `SettingValueSelection.collection_for(name)`

**Review by ezr-ondrej — COMMENTED** [review]
Definitelly make sense, while at it, I would do one more improvement, let me know if that makes sense to you :+1:

**Comment by xprazak2** [comment]
I removed the dymanic method definition.

File: app/models/setting.rb [hunk 3]
```diff
@@ -283,19 +283,22 @@ def self.load_defaults
...
-          collection.is_a?(Hash) ? collection : editable_select_optgroup(collection, :include_blank => options[:include_blank])
-        end
-      end
+      select_collection_registry.add(name, options)
     end
     options[:encrypted] ||= false
     {:name => name, :value => value, :description => description, :default => default, :full_name => full_name, :encrypted => options[:encrypted]}
   end
 
+  def select_collection
```
Comment (ezr-ondrej) [inline app/models/setting.rb:298]: I don't like this to live here, but I can deal with that.

File: app/models/setting.rb [hunk 4]
```diff
@@ -283,19 +283,22 @@ def self.load_defaults
     true
   end
 
+  def self.select_collection_registry
+    @@select_collection ||= SettingSelectCollection.new
```
Comment (ezr-ondrej) [inline app/models/setting.rb:287]: Is there a reason for a class variable instead of instance variable on class object?

**Review by ezr-ondrej — APPROVED** [review]
I have nitpicks, but it's definitelly much better than what we have now :+1:

**Comment by ezr-ondrej** [comment]
[test katello]

**Comment by ezr-ondrej** [comment]
Looks good and I've found only katello tests are using the old helper explicitly (fixed in: https://github.com/Katello/katello/pull/8740), so lets do it and see :)
Thanks @xprazak2 ! :+1:

**Pull request merged**